<a href="https://colab.research.google.com/github/khusanovich/modul_3/blob/main/3_2_5_groupby().ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![Imgur](https://i.imgur.com/5pXzCIu.png)

# Data Science va Sun'iy Intellekt Praktikum

## 3-MODUL. MA’LUMOTLARGA ISHLOV BERISH

### 3.2 Ma'lumotlarni tozalash va tayyorlash

## Guruhlash. `.groupby()`
`.groupby()` meotdi yordamida jadvaldagi ma'lumotlarni biror ustun qiymatlari guruhlarga ajratib tahlil qilishimiz mumkin.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("https://github.com/anvarnarz/praktikum_datasets/raw/main/automobile_data.csv", index_col=0)
df.head()

company   body-style  ...  average-mileage    price
index                            ...                          
0      alfa-romero  convertible  ...               21  13495.0
1      alfa-romero  convertible  ...               21  16500.0
2      alfa-romero    hatchback  ...               19  16500.0
3             audi        sedan  ...               24  13950.0
4             audi        sedan  ...               18  17450.0

[5 rows x 9 columns]

Yuqoridagi jadvalda turli avtomobil brandlari haqida ma'lumotlar jamlangan. Keling boshlanishiga har bir brand uchun ustunlarning o'rtacha qiymatini topamiz:

In [3]:
df.groupby('company').mean()

wheel-base      length  ...  average-mileage         price
company                                ...                               
alfa-romero     90.566667  169.600000  ...        20.333333  15498.333333
audi           101.200000  180.800000  ...        20.000000  16392.500000
bmw            103.433333  185.033333  ...        19.000000  27213.333333
chevrolet       92.466667  151.933333  ...        41.000000   6007.000000
dodge           93.700000  157.300000  ...        31.000000   6303.000000
honda           96.500000  167.200000  ...        26.333333  10195.000000
isuzu           94.433333  160.833333  ...        33.333333   6785.000000
jaguar         109.333333  196.966667  ...        14.333333  34600.000000
mazda           95.900000  164.260000  ...        28.000000   9654.800000
mercedes-benz  113.225000  197.275000  ...        18.000000  35040.000000
mitsubishi      95.000000  164.850000  ...        29.500000   6689.000000
nissan          95.680000  170.140000  ...        31.400000   8289.000000
porsche         92.466667  171.166667  ...        17.000000  35528.000000
toyota          96.957143  167.571429  ...        28.714286   8216.857143
volkswagen      97.300000  171.700000  ...        31.750000   8435.000000
volvo          104.300000  188.800000  ...        23.000000  13177.500000

[16 rows x 5 columns]

Yuqoridagi natijani tahlil qilsak, `alfa-romero` mashinalarning o'rtacha narxi (`price`) `15498$`, `honda` mashinalarining o'rtacha ot kuchi (`horsepower`) `92.3` ga teng ekan.

Bitta ustun uchun qiymatlarni olish ham mumkin:

In [5]:
df.groupby('company').price.max()

company
alfa-romero      16500.0
audi             18920.0
bmw              41315.0
chevrolet         6575.0
dodge             6377.0
honda            12945.0
isuzu             6785.0
jaguar           36000.0
mazda            18344.0
mercedes-benz    45400.0
mitsubishi        8189.0
nissan           13499.0
porsche          37028.0
toyota           15750.0
volkswagen        9995.0
volvo            13415.0
Name: price, dtype: float64

Guruhlashda bir emas, bir nechta ustunlarni ham ko'rsatish mumkin:

In [6]:
df.groupby(['company','body-style']).mean()

wheel-base  ...         price
company       body-style               ...              
alfa-romero   convertible   88.600000  ...  14997.500000
              hatchback     94.500000  ...  16500.000000
audi          sedan         99.666667  ...  15550.000000
              wagon        105.800000  ...  18920.000000
bmw           sedan        103.433333  ...  27213.333333
chevrolet     hatchback     91.450000  ...   5723.000000
              sedan         94.500000  ...   6575.000000
dodge         hatchback     93.700000  ...   6303.000000
honda         sedan         96.500000  ...  11645.000000
              wagon         96.500000  ...   7295.000000
isuzu         sedan         94.433333  ...   6785.000000
jaguar        sedan        109.333333  ...  34600.000000
mazda         hatchback     93.650000  ...   7482.500000
              sedan        104.900000  ...  18344.000000
mercedes-benz hardtop      112.000000  ...  45400.000000
              sedan        115.450000  ...  33256.000000
              wagon        110.000000  ...  28248.000000
mitsubishi    hatchback     93.700000  ...   5789.000000
              sedan         96.300000  ...   7589.000000
nissan        sedan         95.975000  ...   8524.000000
              wagon         94.500000  ...   7349.000000
porsche       convertible   89.500000  ...  37028.000000
              hardtop       89.500000  ...  34028.000000
              hatchback     98.400000  ...           NaN
toyota        hatchback     95.700000  ...   6058.000000
              wagon         97.900000  ...   9836.000000
volkswagen    sedan         97.300000  ...   8435.000000
volvo         sedan        104.300000  ...  12940.000000
              wagon        104.300000  ...  13415.000000

[29 rows x 5 columns]

Jadvaldan xulosa qilsak, `volvo` firmasinig `sedan` turidagi mashinalarining o'rtacha qiymati 12940$, `wagon` turidagi mashinalarning o'rtacha narxi esa 13415$.

`.groupby()` yordamida nafaqat o'rta balki boshqa qiymatlarni ham hisoblash mumkin:

In [9]:
df.groupby(['company']).max()

body-style  wheel-base  ...  average-mileage    price
company                               ...                          
alfa-romero    hatchback        94.5  ...               21  16500.0
audi               wagon       105.8  ...               24  18920.0
bmw                sedan       110.0  ...               23  41315.0
chevrolet          sedan        94.5  ...               47   6575.0
dodge          hatchback        93.7  ...               31   6377.0
honda              wagon        96.5  ...               30  12945.0
isuzu              sedan        94.5  ...               38   6785.0
jaguar             sedan       113.0  ...               15  36000.0
mazda              sedan       104.9  ...               31  18344.0
mercedes-benz      wagon       120.9  ...               22  45400.0
mitsubishi         sedan        96.3  ...               37   8189.0
nissan             wagon       100.4  ...               45  13499.0
porsche        hatchback        98.4  ...               17  37028.0
toyota             wagon       104.5  ...               35  15750.0
volkswagen         sedan        97.3  ...               37   9995.0
volvo              wagon       104.3  ...               23  13415.0

[16 rows x 8 columns]

`.agg()` metodi yordamida esa birdaniga bir nechta statistik ma'lumotlari olishimiz ham mumkin:

In [13]:
df.groupby('company').agg([len, min, max])

body-style                          ... price                  
                     len          min        max  ...   len      min      max
company                                           ...                        
alfa-romero            3  convertible  hatchback  ...   3.0  13495.0  16500.0
audi                   4        sedan      wagon  ...   4.0  13950.0  18920.0
bmw                    6        sedan      sedan  ...   6.0  16430.0  41315.0
chevrolet              3    hatchback      sedan  ...   3.0   5151.0   6575.0
dodge                  2    hatchback  hatchback  ...   2.0   6229.0   6377.0
honda                  3        sedan      wagon  ...   3.0   7295.0  12945.0
isuzu                  3        sedan      sedan  ...   3.0   6785.0   6785.0
jaguar                 3        sedan      sedan  ...   3.0  32250.0  36000.0
mazda                  5    hatchback      sedan  ...   5.0   5195.0  18344.0
mercedes-benz          4      hardtop      wagon  ...   4.0  25552.0  45400.0
mitsubishi             4    hatchback      sedan  ...   4.0   5389.0   8189.0
nissan                 5        sedan      wagon  ...   5.0   6649.0  13499.0
porsche                3  convertible  hatchback  ...   3.0  34028.0  37028.0
toyota                 7    hatchback      wagon  ...   7.0   5348.0  15750.0
volkswagen             4        sedan      sedan  ...   4.0   7775.0   9995.0
volvo                  2        sedan      wagon  ...   2.0  12940.0  13415.0

[16 rows x 24 columns]

`.agg()` yordamida istalgan funksiyalarni chaqirishimiz mumkin, misol uchun `numpy` dagi `mean` (o'rta qiymat) funiksiyasini ham qo'shamiz. Natijalarni faqat `price` ustuni uchun chiqaramiz:

In [14]:
df.groupby('body-style').price.agg([len, min, max, np.mean])

len      min      max          mean
body-style                                       
convertible   3.0  13495.0  37028.0  22341.000000
hardtop       2.0  34028.0  45400.0  39714.000000
hatchback    15.0   5151.0  16500.0   7159.571429
sedan        32.0   6575.0  41315.0  17706.333333
wagon         9.0   6918.0  28248.0  12730.111111

`.apply()` metodi yordamida guruhdagi qiymatlarni o'zgartirishimiz ham mumkin: